Fix #6049: Solana Dapps Javascript communication layer

StephenHeaps opened the pull request:
## Summary of Changes
- PR is intended for initial security review of the javascript communication layer for Solana dapps.

Some features are currently unavailable:
- Creating a solana account automatically if one does not yet exist when connecting to a Solana dapp: https://github.com/brave/brave-ios/issues/6046
- Solana dapp transaction parsing (transactions will appear with some incorrect / unclear info): https://github.com/brave/brave-ios/issues/6004
- Solana Sign Transaction / Sign All Transactions panel: https://github.com/brave/brave-ios/issues/6005
- Connected button UI on wallet panel: https://github.com/brave/brave-ios/issues/5943 & https://github.com/brave/brave-ios/issues/5944


This pull request fixes #6049

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Test Plan:
1. Enable `isSolanaDappsEnabled` feature flag in `WalletConstants.swift`
2. Build & run the project, visit Solana dapps websites.
    - Brave test site [hosted](https://codesandbox.io/s/github/darkdh/solana-provider-test) ([source](https://github.com/darkdh/solana-provider-test))
    - [MagicEden](https://magiceden.io/)
    - [Jupiter](https://jup.ag/)


## Screenshots:

https://user-images.githubusercontent.com/5314553/191633069-525fb167-55cc-4c2d-a9ef-8b59dade9825.mov


## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

StephenHeaps committed:
Add WalletSolanaProvider.js script, injected into each tab with the Solana Provider js from BraveCore in `UserScriptManager`.
Add base for `SolanaProviderHelper` as a `TabContentScript`, add it as a content script.
Add `accountChanged` event to `Web3ProviderEvent` constants.
Add `Preferences.Wallet.defaultSolWallet` for tracking the default wallet used for web3 Solana apis.

StephenHeaps committed:
Update WalletPanelView to check for pending requests for Ethereum or Solana coin types.

StephenHeaps committed:
Lazy-load the Solana internal script when `createPublicKey` / `createTransaction` are needed (performance reasons).

StephenHeaps committed:
Add missing `BraveWalletProviderDelegate` functions for Solana connected accounts

StephenHeaps committed:
Add separate preference for Ethereum/Solana provider access. Replace existing `Allow Sites to Request Accounts`.

StephenHeaps committed:
`BraveWalletProviderDelegate` updates for Solana connected accounts

nuo-xu committed:
fix rebasing error.

StephenHeaps committed:
Convert Swift dictionary to JS for created public key

StephenHeaps committed:
Don't inject Solana internal script in `updateSolanaProperties()`. Only should be injected when receiving a `connect` request, or creating a transaction.

StephenHeaps committed:
Sandbox the Solana internal script

StephenHeaps committed:
Fix connect request removing `PublicKey` from object, fix replying with optional `createdPublicKey` string value.

StephenHeaps committed:
Fix for $Object with Core 1.45.x.
Fix connect event by interacting with `solanaWeb3.PublicKey` within JS so we maintain the object type & it's methods.

StephenHeaps committed:
Fix signMessage request

StephenHeaps committed:
Fix connect/disconnect flow from request

StephenHeaps committed:
Comments for expected Solana dapp request response values

StephenHeaps committed:
Fix solana internal script injection. Use multi-line so comments in the internal script don't comment out the trailing braces of our function.

StephenHeaps committed:
SignAndSendTransaction

StephenHeaps committed:
Add work-in-progress `.solDappTransaction` to TransactionParser` so the solana dapp transactions can be rejected individually

StephenHeaps committed:
Add base `SignTransactionView`, support displaying sign transaction & sign all transactions requests

StephenHeaps committed:
Sign Transaction

StephenHeaps committed:
Sign All Transactions

StephenHeaps committed:
Fix error message response, now formatted as a dictionary/object with the error code.

StephenHeaps committed:
Fix `signMessage`, `signAndSendTransaction` success request response

StephenHeaps committed:
Request success response

StephenHeaps committed:
Reduce javascript

StephenHeaps committed:
Fix encoded result for successful `signAndSendTransaction` transactions, and `request` requests.

StephenHeaps committed:
Rename `PresentingContext.requestEthereumPermissions` to `PresentingContext.requestPermissions` as any coin type is supported

StephenHeaps committed:
Fixed accountChanged event not emitting solanaWeb3.PublicKey object

StephenHeaps committed:
Inject internal script when receiving connect request, before calling `connect()` on the `BraveWalletSolanaProvider`.
Update `window.solana.publicKey` assignment function call.

kylehickinson committed:
Fix memory leak when showing a wallet notification

StephenHeaps committed:
Clear solana connected accounts in `func webView(_:didCommit:)` for the tab.

StephenHeaps committed:
Call `updateSolanaProperties()` before replying to `connect` request so that `window.solana.publicKey` is assigned and add a check for `_brave_solana.createPublickey` exists before calling it.

StephenHeaps committed:
Only assign `window.solana.publicKey` when the account is connected

StephenHeaps committed:
Fix parsing issues for `signMessage` requests via `window.solana.request`.

StephenHeaps committed:
Cleanup, add `Keys` constants, `MojoBase.Value` helpers.

StephenHeaps committed:
Move solana dapps integration logic behind feature flag

StephenHeaps committed:
Move SolanaProviderHelper requests to their own functions

StephenHeaps committed:
Restore solana dapps feature flag to disabled

StephenHeaps committed:
Fix script injection when debug flag disabled

StephenHeaps committed:
Fix connecting second Solana account after one previously connected to dapp site

StephenHeaps committed:
Show Solana accounts in `EditSiteConnectionView` when the selected account is a Solana account.

StephenHeaps committed:
Fix account changed event emit when switching to a non-permitted account.

StephenHeaps committed:
Add `SolanaProviderHelperTests` for testing the responses from each Solana dapp request for error and success flows. Add comments to `SolanaProviderHelper` functions.

kdenhartog reviewed (commented):
initial review. So far things are looking pretty good

StephenHeaps committed:
Address PR feedback, rename `function convertSignaturePubkeyPair(signaturePubkeyPair)` -> `function convertSignaturePubkeyTuple(signaturePubkeyTuple)`

StephenHeaps committed:
Cleanup request permissions flow for 1 solana account, fix connect internal error missing error message.

StephenHeaps committed:
Updates after JS Refactor PR merge & rebase.

StephenHeaps committed:
Resolve tests for compile after rebase & updates.

StephenHeaps committed:
Inject solanaWeb3 library, call `solanaWeb3.PublicKey` & `solanaWeb3.Transaction` constructors directly instead of through the global functions on `_brave_solana`. Remove `_brave_solana` property definition.

kdenhartog reviewed (approved):
I think this one is good to move forward for now as mentioned in DM. We've got a secondary issue https://github.com/brave/brave-ios/issues/6154 that will double check all this has been closed out properly and this will let the front end work progress forward so we can still hopefully get this into 1.45.x iOS release. See the [security issue](https://github.com/brave/security/issues/1040) for this for additional details.

StephenHeaps committed:
Address PR feedback, feature flag `EditSiteConnectionView` to use ethereum coin type when `isSolanaDappsEnabled` flag is false.

iccub reviewed (commented):
code looks good to me. only concerns i had were about public address sanitization in some methods

StephenHeaps committed:
Update `SolanaProviderScriptHandler` for logger changes

StephenHeaps committed:
Address PR feedback. Localize `SignTransactionView` strings, sanitize public key strings before executing javascript for `connect` event, `accountChanged` event and assigning `window.solana.publicKey`.

StephenHeaps committed:
Address PR feedback.
Remove redundant `@MainActor` from Tasks when Task is only calling a functionalready marked `@MainActor`.
Don't wrap entire function in `SolanaProviderScriptHandler` with `Task`, only the code requiring async context.
Remove unused `shouldInjectWalletProvider(_:)` functions from `SolanaProviderScriptHandler` and `EthereumProviderScriptHandler`. Both were unused, and checking for default keyring exists but we have an onboarding flow `showWalletOnboarding()` in `BraveWalletProviderDelegate`.
